### ISSUE TITLE
ndk-r15c compile error

### ISSUE THREAD

**Issue by qtfreet00** [smartsheet-mobile/smjni#1]
```
smjni/inc\smjni/java_ref.h:54:18: error: field has incomplete type 'smjni::java_ref<_jstring *, smjni::internal::auto_ref_traits>'
        java_ref(nullptr_t) noexcept:
                 ^
smjni/inc\smjni/java_string.h:44:13: note: in instantiation of template class 'smjni::java_ref<_jstring *, smjni::internal::auto_ref_traits>' requested here
        if (!str)
            ^
smjni/inc\smjni/java_ref.h:41:11: note: definition of 'smjni::java_ref<_jstring *, smjni::internal::auto_ref_traits>' is not complete until the closing '}'
    class java_ref : private Traits
          ^
In file included from smjni/src/java_string.cpp:18:
smjni/inc\smjni/java_string.h:44:13: error: invalid argument type 'const auto_java_ref<jstring>' (aka 'const smjni::java_ref<_jstring *, smjni::internal::auto_ref_traits>') to unary expression
        if (!str)
            ^~~~
smjni/inc\smjni/java_string.h:46:46: error: no matching member function for call to 'c_ptr'
        jsize ret = env->GetStringLength(str.c_ptr());
                                         ~~~~^~~~~

```

**Comment by gershnik** [comment]
@qtfreet00 Thank you for reporting! Should be good now.

**Comment by qtfreet00** [comment]
a new error occurred

my environment
```
Android Studio 3.5
Ndk r15c
Windows 10
```

```
In file included from app/src/main/cpp/smjni/inc\smjni/smjni.h:30:
app/src/main/cpp/smjni/inc\smjni/java_array.h:55:5: error: C++ requires a type specifier for all declarations
    java_array_access(JNIEnv * env, T array) -> java_array_access<T, IsObject>;
    ^
app/src/main/cpp/smjni/inc\smjni/java_array.h:58:5: error: C++ requires a type specifier for all declarations
    java_array_access(JNIEnv * env, const java_ref<T, Traits> & array) -> java_array_access<T, IsObject>;
    ^
app/src/main/cpp/smjni/inc\smjni/java_array.h:422:9: error: use of class template 'java_array_access' requires template arguments
        java_array_access res_access(env, res.c_ptr());
        ^
app/src/main/cpp/smjni/inc\smjni/java_array.h:52:11: note: template is declared here
    class java_array_access;
          ^
3 errors generated.
ninja: build stopped: subcommand failed.
```

**Comment by ghost** [comment]
Ah, this is C+17 feature actually. Does NDK 15 compiler support C++ 17? If you are using our CMakeLists.txt you can quickly check by setting CXX_STANDARD to 17 instead of 14 there and see if everything builds fine. 
If C++17 isn't well supported we can #ifdef unsupported features out for older compilers.

**Comment by qtfreet00** [comment]
yes, ndk 15 does not support c++17, I tested on ndk 20 ,it works

but for some reason, I need to use ndk 15 instead

**Comment by ghost** [comment]
No problem. I am going to #ifdef couple of things out.

**Comment by ghost** [comment]
Pull and try now.

**Comment by qtfreet00** [comment]
it works , thank you☺

**Comment by ghost** [comment]
You are welcome!
Note that since you are using vnext branch which is not yet documented on the wiki some info there might be out of date. 
Also the code generator in jnigen directory isn't documented yet (but functional)

**Comment by qtfreet00** [comment]
yes ,There are many differences with branch master

**Comment by gershnik** [comment]
@qtfreet00 the next branch had been merged to master and became version 2.0. The wiki had been updated to reflect changes so (hopefully) everything should be nice and tidy now. :)
Hope it works for you!

**Comment by qtfreet00** [comment]
I'll try it. Thank you for your work. ☺
